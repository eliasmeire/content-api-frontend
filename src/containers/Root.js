import React from 'react';
import {Provider} from 'react-redux';
import App from './App';

const Root = ({store, actions}) => (
  <Provider store={store}>
    <App actions={actions} />
  </Provider>
);

export default Root;