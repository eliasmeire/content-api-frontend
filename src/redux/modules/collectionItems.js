import { getCollectionItems, getItem } from '../../utils/apiClient';
import jwt from '../../utils/jwt';
import 'rxjs';
import { Observable } from 'rxjs/Observable';

export const initialState = {
  error: null,
  list: [],
  selected: {}
};

export const reducer = (state = initialState, action) => {
  switch(action.type) {
    case 'FETCH_ITEMS_FULFILLED':
      return {...state, error: null, list: action.items};
    case 'FETCH_ITEMS_REJECTED':
      return {...state, error: action.error};
    case 'FETCH_ITEM_FULFILLED':
      return {...state, error: null, selected: action.item};
    case 'FETCH_ITEM_REJECTED':
      return {...state, error: action.error};
    case 'DELETE_ITEMS':
      return initialState;
    default:
      return state;
  }
}

//actions
export const fetchItems = collName => ({type: 'FETCH_ITEMS', collName});
export const fetchItemsFulfilled = items => ({type: 'FETCH_ITEMS_FULFILLED', items});
export const fetchItemsRejected = error => ({type: 'FETCH_ITEMS_REJECTED', error});
export const fetchItem = (collName, id) => ({type: 'FETCH_ITEM', id, collName});
export const fetchItemFulfilled = item => ({type: 'FETCH_ITEM_FULFILLED', item});
export const fetchItemRejected = error => ({type: 'FETCH_ITEM_REJECTED', error});
export const deleteItems = () => ({type: 'DELETE_ITEMS'});

//epics
export const fetchItemsEpic = action$ =>
  action$.ofType('FETCH_ITEMS')
    .mergeMap(action =>
      getCollectionItems(jwt.get(), action.collName)
        .map(res => res.response.data)
        .map(items => fetchItemsFulfilled(items))
        .catch(err => Observable.of(fetchItemsRejected(err)))
    );

export const fetchItemEpic = action$ =>
  action$.ofType('FETCH_ITEM')
    .mergeMap(action =>
      getItem(jwt.get(), action.collName, action.id)
        .map(res => res.response.data)
        .map(item => fetchItemFulfilled(item))
        .catch(err => Observable.of(fetchItemRejected(err)))
    );
