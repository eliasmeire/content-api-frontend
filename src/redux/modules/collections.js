import { getCollections } from '../../utils/apiClient';
import jwt from '../../utils/jwt';
import 'rxjs';
import { Observable } from 'rxjs/Observable';

export const initialState = {
  error: null,
  data: []
};

export const reducer = (state = initialState, action) => {
  switch(action.type) {
    case 'FETCH_COLLECTIONS_FULFILLED':
      return {error: null, data: action.collections};
    case 'FETCH_COLLECTIONS_REJECTED':
      return {...state, error: action.error};
    case 'DELETE_COLLECTIONS':
      return initialState;
    default:
      return state;
  }
}

//actions
export const fetchCollections = () => ({type: 'FETCH_COLLECTIONS'});
export const fetchCollectionsFulfilled = collections => ({type: 'FETCH_COLLECTIONS_FULFILLED', collections});
export const fetchCollectionsRejected = error => ({type: 'FETCH_COLLECTIONS_REJECTED', error});
export const deleteCollections = () => ({type: 'DELETE_COLLECTIONS'});

//epics
export const fetchCollectionsEpic = action$ =>
  action$.ofType('FETCH_COLLECTIONS')
    .mergeMap(action => 
      getCollections(jwt.get())
        .map(res => res.response.data)
        .map(data => fetchCollectionsFulfilled(data))
        .catch(err => Observable.of(fetchCollectionsRejected(err)))
    );  