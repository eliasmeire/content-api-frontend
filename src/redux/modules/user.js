import { auth } from '../../utils/apiClient';
import jwt from '../../utils/jwt';

export const initialState = {
  prevUrl: '/',
  loggedIn: false,
  fetching: false,
  user: {
    avatar_url: null,
    username: null,
    id: null,
    permissions: {}
  }
};

export const reducer = (state = initialState, action) => {
  switch(action.type) {
    case 'LOGIN_PENDING':
      return {...state, fetching: true};
    case 'LOGIN_FULFILLED':
      return {...state, fetching: false, loggedIn: true, user: action.user};
    case 'LOGIN_REJECTED':
      return {...state, fetching: false};
    case 'LOGOUT':
      return {...state, loggedIn: false, user: {}};
    case 'SET_PREV_URL':
      return {...state, prevUrl: action.url};
    case 'LOAD_TOKEN_FULFILLED':
      return {...state, loggedIn: true, user: action.user}
    default:
      return state;
  }
}

// actions
export const loginPending = () => ({ type: 'LOGIN_PENDING'});
export const loginRejected = () => ({ type: 'LOGIN_REJECTED'});
export const loginFulfilled = (user) => ({ type: 'LOGIN_FULFILLED', user});
export const setPrevUrl = (url) => ({type: 'SET_PREV_URL', url});
export const loadTokenFulfilled = (user) => ({type: 'LOAD_TOKEN_FULFILLED', user});
export const logout = () => {
  jwt.remove();
  return { type: 'LOGOUT' };
};

export const loadToken = () => {
  return (dispatch) => {
    const token = jwt.getDecoded();
    console.log(token)
    if(token) {
      dispatch(loadTokenFulfilled(token.data));
      return true;
    }
    return false;
  };
};

export const login = (user) => {
  return (dispatch) => {
    dispatch(loginPending());
    return new Promise((resolve, reject) => {
      auth(user)
        .then((res) => {
          setTimeout(1000);
          if(res.status !== 200) {
            dispatch(loginRejected());
            reject();            
          }
          return res.json();
        })
        .then((json) => {
          let token = json.data;
          if(token) {
            jwt.store(token)
            let decoded = jwt.decode(token);
            dispatch(loginFulfilled(decoded.data));
            resolve()
          }          
        })
        .catch((err) => {
          dispatch(loginRejected());
          reject();
        });
    }); 
  };
};