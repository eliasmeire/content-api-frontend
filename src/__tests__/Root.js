import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { initialState as user } from '../redux/modules/user';
import Root from '../containers/Root';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('<Root />', () => {
  it('renders with default props', () => {
    const store = mockStore({ user });
    const wrapper = shallow(
      <Root store={store} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
