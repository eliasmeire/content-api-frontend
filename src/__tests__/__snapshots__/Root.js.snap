exports[`<Root /> renders with default props 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }>
  <Connect(App) />
</Provider>
`;
